pdns/misc.cc

```diff
@@ -1745,3 +1740,26 @@ bool constantTimeStringEquals(const std::string& a, const std::string& b)
 #endif /* !HAVE_SODIUM_MEMCMP */
 #endif /* !HAVE_CRYPTO_MEMCMP */
 }
+
+namespace pdns
+{
+std::optional<std::string> visit_directory(const std::string& directory, const std::function<bool(ino_t inodeNumber, const std::string_view& name)>& visitor)
+{
+  auto dirHandle = std::unique_ptr<DIR, decltype(&closedir)>(opendir(directory.c_str()), closedir);
+  if (!dirHandle) {
+    auto err = errno;
+    return std::string("Error opening directory '" + directory + "': " + stringerror(err));
```

Comment:
I would prefer to throw an exception with an object that contains the error message extracted from `errno` using `pdns::getMessageFromErrno()` directly.

Reply:
I find it hard to make the right choice between returning a potential error or throwing an exception. In that case I opted for the former because most callers seem to expect the directory not to exist, or not to be readable, so I felt that returning the error was better than forcing the caller to wrap the call in a `try .. catch` block, but I think it really is a matter of taste.